libkbfs/block_journal.go

```diff
@@ -1105,6 +1138,95 @@ func (j *blockJournal) ignoreBlocksAndMDRevMarkers(ctx context.Context,
 	return nil
 }
 
+func (j *blockJournal) saveBlocksUntilNextMDFlush() error {
+	// Copy the current journal entries into a new journal.  After the
+	// next MD flush, we can use the saved journal to delete the block
+	// data for all the entries in the saved journal.
+	first, err := j.j.readEarliestOrdinal()
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	last, err := j.j.readLatestOrdinal()
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	savedJournal := j.saveUntilMDFlush
+	if savedJournal == nil {
+		savedJournalDir, err := ioutil.TempDir(j.dir, "saved_block_journal")
+		if err != nil {
+			return err
+		}
+
+		sj := makeDiskJournal(
+			j.codec, savedJournalDir, reflect.TypeOf(blockJournalEntry{}))
+		savedJournal = &sj
+	}
+
+	for i := first; i <= last; i++ {
+		e, err := j.readJournalEntry(i)
+		if err != nil {
+			return err
+		}
+
+		savedJournal.appendJournalEntry(nil, e)
+	}
+
+	j.saveUntilMDFlush = savedJournal
+	return nil
+}
+
+func (j *blockJournal) onMDFlush() error {
+	if j.saveUntilMDFlush == nil {
+		return nil
+	}
+
+	// Delete the block data for anything in the saved journal.
+	first, err := j.saveUntilMDFlush.readEarliestOrdinal()
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	last, err := j.saveUntilMDFlush.readLatestOrdinal()
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	for i := first; i <= last; i++ {
+		e, err := j.saveUntilMDFlush.readJournalEntry(i)
+		if err != nil {
+			return err
+		}
+
+		entry, ok := e.(blockJournalEntry)
```

Comment:
Done, though it seems a little weird to me now, since the earliest ordinal is getting read for disk twice now on the normal path.

Note I left `entry` as a param to avoid reading it from disk twice in the happy path.

Is that what you were thinking?